### PR TITLE
Addison Cole - Edges - Grocery Store

### PR DESCRIPTION
**Pull request by addisoncole**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

Question | Response
--- | ---
What is accomplished with `raise ArgumentError`? | It is a way to raise an error by the programmer when there might not be a more specific exception raised by the program
Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods? |Because they are methods that deal with all instances of Customer or all instances of Order, rather than a single instance or object of Order or Customer. 
Think about the relation between `Order` and `Customer`. Is this relation one-to-one, one-to-many, or something else? How does that compare to the Solar System project? | It is One to Many as each customer has many orders, however, there are may customers with many orders. 
How is the relation between `Order` and `Customer` tracked in the CSV file? How is it tracked in your program? Why might these be different? | The only data point the two CSV's share is the Customer ID. In my program I accessed the CSV info of each customer and made them an instance so that an ID was associated with a specific customer, then doing the same for orders (access all orders and assign an object ID for each), I can connect the dots between the Customer ID for the object of Customer and their ID in the customer ID portion of the Order.
Did the presence of automated tests change the way you thought about the problem? How? | It allowed me to slow down and think about each method in its most simplest forms, work to achieve those, and then start thinking about edge cases and ensuring the method works correctly.
CS Fun: If there are `C` customers listed in `customers.csv` and `R` orders listed in `orders.csv`, what is the time complexity of calling `Customer.find`? What is the time complexity of `Order.find`? |C^n is the complexity of calling Customer.find. It is a linear time complexity.

### CODE REVIEW

**Comment by tildeee** [comment]
# Grocery Store
## What We're Looking For

Feature | Feedback
--- | ---
**Baseline** |
Answered comprehension questions | x, what about big o for `Order.find`?
Used Git Regularly | x
**Wave 1** |
All provided tests pass | x
Using the appropriate `attr_` for instance variables | x
**Wave 2** |
All stubbed tests are implemented fully and pass | x
Used `CSV` library only in `.all` (not in `.find`) | x
Appropriately parses the product data from CSV file in `Order.all` | x
`Order.all` calls `Customer.find` to set up the composition relation | x
**Additional Notes** | 

Great work Addison! The project looks great. I think your code is neat and well-organized.

In fact, I think most of my comments for improvements would be to encourage you to think about fancier `Enumerable` methods when searching for things (`find`) or summing things (`sum`) as opposed to using `.each`, but this is purely optional.

Also, I think your tests are really nice and concrete.

Well done overall!

File: lib/order.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      # go through CSV product list for the current customer, split each for processing into a hash to meet initialization requirement
+      order_products = {}
+      raw_products = row[1]
+      raw_products = raw_products.split(';')
+      raw_products.each do |raw_product|
+        raw_product = raw_product.split(':')
+        product = { raw_product[0] =>
+          raw_product[1].to_f }
+          order_products = order_products.merge(product)
+        end
```
Comment (tildeee) [inline lib/order.rb:63]: your indentation is a little off in this `.each` block so it confused me for a second!

File: lib/order.rb [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    def self.find_by_customer(customer_id)
+      @customer_orders = []
+      @orders = Order.all
+      customer_id = Order.find(customer_id)
+      @orders.each do |order|
+        if order.customer.id == customer_id.id
+          @customer_orders << order
+        end
+      end
```
Comment (tildeee) [inline lib/order.rb:100]: Great opportunity for a `select`

File: lib/order.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      new_item = {item => price}
+      @products = @products.merge(new_item)
+    end
+  end
+
+  def remove_product(item)
+    if @products.has_key?(item) == false
+      return raise ArgumentError, 'Item not on list'
+    else
+      @products = @products.delete_if { |key, value| key == item }
```
Comment (tildeee) [inline lib/order.rb:36]: nicely done!

File: lib/order.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+require_relative 'customer.rb'
+
+class Order
+
+  attr_reader :id
+  attr_accessor :products, :customer, :fulfillment_status
+
+  FULLFILLMENT_STATUS = [:pending, :paid, :processing, :shipped, :complete]
+
+  def initialize (id, products, customer, status = :pending)
```
Comment (tildeee) [inline lib/order.rb:12]: Glad you used a default value!
